From af260092ea6fb23c61cea1d4b239da0334ad341c Mon Sep 17 00:00:00 2001
From: "H. Vetinari" <h.vetinari@gmx.com>
Date: Fri, 12 Feb 2021 10:57:44 -0800
Subject: [PATCH 08/19] log success messages and errors in loader.py (#1683)

Summary:
This patch is to provide log information in case of failing to load the AVX2-enabled lib,
and also to log the successful loading of the libraries.

Otherwise, the programmatic detection of which library was loaded also needs to take
into account which other log statements were executed, if the original import failed due to an exception,
cf. the tests in https://github.com/conda-forge/faiss-split-feedstock/pull/27

Finally, this reduces the number of import statements by one - one non-AVX2 fallback is enough.

Related to https://github.com/facebookresearch/faiss/issues/1600, https://github.com/facebookresearch/faiss/issues/1680, https://github.com/facebookresearch/faiss/issues/1681, https://github.com/facebookresearch/faiss/issues/1682.

Pull Request resolved: https://github.com/facebookresearch/faiss/pull/1683

Reviewed By: mdouze

Differential Revision: D26424333

Pulled By: beauby

fbshipit-source-id: 16beddec7e0c098b913a7f5420cbb02d1cf515ad
---
 faiss/python/loader.py | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/faiss/python/loader.py b/faiss/python/loader.py
index 49601edf..d8d2216d 100644
--- a/faiss/python/loader.py
+++ b/faiss/python/loader.py
@@ -45,16 +45,19 @@ def supported_instruction_sets():
 
 logger = logging.getLogger(__name__)
 
-try:
-    has_AVX2 = "AVX2" in supported_instruction_sets()
-    if has_AVX2:
+has_AVX2 = "AVX2" in supported_instruction_sets()
+if has_AVX2:
+    try:
         logger.info("Loading faiss with AVX2 support.")
         from .swigfaiss_avx2 import *
-    else:
-        logger.info("Loading faiss.")
-        from .swigfaiss import *
+        logger.info("Successfully loaded faiss with AVX2 support.")
+    except ImportError as e:
+        logger.info(f"Could not load library with AVX2 support due to:\n{e!r}")
+        # reset so that we load without AVX2 below
+        has_AVX2 = False
 
-except ImportError:
+if not has_AVX2:
     # we import * so that the symbol X can be accessed as faiss.X
     logger.info("Loading faiss.")
     from .swigfaiss import *
+    logger.info("Successfully loaded faiss.")
-- 
2.29.2.windows.3

